### PR TITLE
Support CTest

### PR DESCRIPTION
**Pull request by hsorby**
Adding support for running `CTest` as a testing step.

### CODE REVIEW

**Comment by p12tic** [comment]
Looks good, but we need all behavior tested. I'll get back with an example of how this could look like.

**Comment by p12tic** [comment]
@hsorby: A good example of dow to write tests for a feature linke this would be in `master/buildbot/test/unit/test_steps_cppcheck.py`

File: master/buildbot/steps/ctest.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import re
```
Comment (p12tic) [inline master/buildbot/steps/ctest.py]: Licensing comment is missing.

**Comment by p12tic** [comment]
Looks great. Could you rebase the branch and remove unnecessary commits?

File: master/buildbot/steps/ctest.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+class CTestTestCounter(logobserver.LogLineObserver):
+    _summary_line_re = re.compile(r'[0-9]+% tests passed, ([0-9]+) tests failed out of ([0-9]+)')
+    _progress_line_re = re.compile(r'^([0-9]+)/([0-9]+) Test')
+
+    def __init__(self):
+        super().__init__()
+        self._num_tests = 0
+        self._finished = False
+        self._counts = {'total': 0,
```
Comment (sa2ajj) [inline master/buildbot/steps/ctest.py]: could this be re-formatted to make it friendlier for future reviews? e.g.

```!python
self._counts = {
    'total': 0,
    'failures': 0,
}
```

File: master/buildbot/steps/ctest.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        self.progressMetrics += ('tests',)
+
+    def evaluateCommand(self, cmd):
+
+        tests_failed_count = self._observer.failed_count()
+        total_tests = self._observer.total_count()
+
+        self.setTestResults(failed=tests_failed_count, total=total_tests, passed=total_tests - tests_failed_count)
+
+        if tests_failed_count == 0 and total_tests > 0:
```
Comment (sa2ajj) [inline master/buildbot/steps/ctest.py]:
```!python
if not tests_failed_count and total_tests:
```
?

**Pull request closed without merging**